### PR TITLE
ENH: Sorting of ExtensionArrays

### PR DESCRIPTION
**Pull request by TomAugspurger**
This enables {Series,DataFrame}.sort_values and {Series,DataFrame}.argsort. These are required for `factorize`, which is required for `groupby` (my end goal).

I haven't implemented `ExtensionArray.sort_values` yet because it hasn't become necessary. But I can if needed / desired.

### CODE REVIEW

File: pandas/core/arrays/base.py [hunk 1]
```diff
@@ -216,6 +216,27 @@ def isna(self):
...
+
+        Parameters
+        ----------
+        axis : int or None, optional
+            Axis along which to sort. ExtensionArrays are 1-dimensional,
+            so this is only included for compatibility with NumPy.
+        kind : {'quicksort', 'mergesort', 'heapsort'}, optional
+            Sorting algorithm.
+        order : str or list of str, optional
+            Included for NumPy compatibility.
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/base.py]: Is this compatibility needed because in the code we use `np.argsort(values)` which passes those keywords to the method? 
(it is a bit unfortunate ..)

Comment (TomAugspurger) [reply pandas/core/arrays/base.py]: It's not necessary, and I can remove it.

I see now that Categorical.argsort has a different signature. I suppose we should match that.

**Comment by TomAugspurger** [comment]
Changed how this is organized a bit, to reflect a pattern I noticed here an elsewhere.

In several places (here, factorize, unique), a method like `.argsort` is composed of three parts

1.) Data coercion / prep
2.) The actual algorithm
3.) postprocessing

In the case of argsort, it's

1.) Just the array for most types, the codes for `Categorical`
2.) `np.argsort`
3.) Maybe reversiong for ascending=False

So I split the method in two `EA.argsort` and `EA._values_for_argsort`. For the common case of "I just want to pick which array gets sent to the algo", you just have to overrride `_values_for_argort`. If you need total control (e.g. if you aren't using `np.argsort` to do the actual work), then you'll need to override `EA.argsort` and do everything.

I don't know how useful this will prove to be, but wanted to hear other's thoughts.

**Comment by jorisvandenbossche** [comment]
Do you foresee similar patterns for other algos? Like `_values_for_factorize` (not sure if that makes sense). Just to think about if we would get a proliferation of such methods

**Comment by TomAugspurger** [comment]
Yes factorize would be another. Though it would be a bit more complicated.

I'll probably remove it for now. That means a bit more duplication, but
fewer levels of indirection.

**Comment by shoyer** [comment]
What is the use-case for writing your own sorting algorithm? Maybe radix sort when your data falls into pre-known categories?

My inclination would be to only cinlude `_values_for_argsort`, since that leaves extension array authors the minimal work.

**Comment by TomAugspurger** [comment]
> my inclination would be to only cinlude _values_for_argsort

Do you mean not having an `argsort` method then? From pandas point of view, having `argsort`, as we don't have to check the array time in `sort_values`, and other places where we use `arr.argsort`.

File: pandas/core/arrays/categorical.py [hunk 2]
```diff
@@ -1390,17 +1390,24 @@ def check_for_ordered(self, op):
...
-        compatibility with numpy ndarray objects.
+    def _values_for_argsort(self):
+        return self._codes.copy()
 
-        While an ordering is applied to the category values, arg-sorting
-        in this context refers more to organizing and grouping together
-        based on matching category values. Thus, this function can be
-        called on an unordered Categorical instance unlike the functions
-        'Categorical.min' and 'Categorical.max'.
+    def argsort(self, *args, **kwargs):
```
Comment (TomAugspurger) [inline pandas/core/arrays/categorical.py]: Not sure if this changes our opinion on `_values_for_argsort`, but the apparently Python2 has issues with passing through the arguments correctly to the `super()` call.

```
____________________ TestCategoricalSort.test_numpy_argsort ____________________

self = <pandas.tests.categorical.test_sorting.TestCategoricalSort object at 0x7efcb391f950>

    def test_numpy_argsort(self):
        c = Categorical([5, 3, 1, 4, 2], ordered=True)
    
        expected = np.array([2, 4, 1, 3, 0])
>       tm.assert_numpy_array_equal(np.argsort(c), expected,
                                    check_dtype=False)

pandas/tests/categorical/test_sorting.py:26: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../miniconda3/envs/pandas/lib/python2.7/site-packages/numpy/core/fromnumeric.py:886: in argsort
    return argsort(axis, kind, order)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = [5, 3, 1, 4, 2]
Categories (5, int64): [1 < 2 < 3 < 4 < 5]
ascending = -1, kind = 'quicksort', args = (None,), kwargs = {}

    def argsort(self, ascending=True, kind='quicksort', *args, **kwargs):
        """
            Returns the indices that would sort the Categorical instance if
            'sort_values' was called. This function is implemented to provide
            compatibility with numpy ndarray objects.
    
            While an ordering is applied to the category values, arg-sorting
            in this context refers more to organizing and grouping together
            based on matching category values. Thus, this function can be
            called on an unordered Categorical instance unlike the functions
            'Categorical.min' and 'Categorical.max'.
    
            Returns
            -------
            argsorted : numpy array
    
            See also
            --------
            numpy.ndarray.argsort
            """
        # Keep the implementation here just for the docstring.
        return super(Categorical, self).argsort(ascending=ascending, kind=kind,
>                                               *args, **kwargs)
E       TypeError: argsort() got multiple values for keyword argument 'ascending'
```

Changing the Categorical.argsort to accept just `*args, **kwargs` fixes things, since `ExtensionArray` does the argument validation, but it's a bit unfortunate.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/base.py [hunk 3]
```diff
@@ -236,6 +237,52 @@ def isna(self):
         """
         raise AbstractMethodError(self)
 
+    def _values_for_argsort(self):
+        # type: () -> ndarray
+        """Get the ndarray to be passed to np.argsort.
```
[the remark this answers is not in the document]
Comment (TomAugspurger) [reply pandas/core/arrays/base.py]: `_ndarray_valaues` seems like more of an internal thing, no? I don't know enough to say whether the `_ndarray_values` appropriate for our current uses (mostly indexing IIRC) are also appropriate for argsort.

[the remark this answers is not in the document]
Comment (jreback) [reply pandas/core/arrays/base.py]: my point is what is the point of overriding this specific one? why is not a general purpose EA method/property used here. The proliferation of methods properties is really troublesome.

[the remark this answers is not in the document]
Comment (jreback) [reply pandas/core/arrays/base.py]: if someone wants to override argsort great. but providing an indirect mechism is really kludgey.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (TomAugspurger) [reply pandas/core/arrays/base.py]: > my point is what is the point of overriding this specific one?

This PR is implementing argsort. I could see a similar pattern for factorize.

> The proliferation of methods properties is really troublesome.

How so?

> but providing an indirect mechism is really kludgey.

What's kudgey about it? I think the common case will be overriding the values provided to `np.argsort`, not overriding the sorting algorithm itself. This is true for Categorical and IPArray, and will be true for Period and probably others.

Without `_values_for_argsort` we, and 3rd party libraries, will have duplicate code for validating keyword arguments and the ascending kwarg.

[the remark this answers is not in the document]
Comment (jreback) [reply pandas/core/arrays/base.py]: why is this different from ``_ndarray_values``, which is implemented on EA arrays? having 2 is just plain confusing, not to mention you have a ``_formatting_values```.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (TomAugspurger) [reply pandas/core/arrays/base.py]: FYI, `_ndarray_values` is present on *containers* like Series / Index. We don't have a similar convention for the arrays themselves.

`_formatting_values` is specific for printing, which we certainly don't want here.

I'm going to do something like this when moving Period to an EA, otherwise we'll have duplicate logic for parameter validation and the ascending in 3 places. It's really just a matter of whether we want to expose it to library authors implementing an EA.

**Comment by TomAugspurger** [comment]
The linting failure is fixed in https://github.com/pandas-dev/pandas/pull/20330

File: pandas/core/arrays/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jreback) [reply pandas/core/arrays/base.py]: my point is that I don't want a specific ``_values_for_argsort`` but rather something *like* ``_ndarray_values``, and I would just call it that. you need to turn this into an ndarray sometimes, so provide a methdo/property for that.
OR simply call ``np.asarray()`` if you actually need an ndarray.

File: pandas/core/arrays/categorical.py [hunk 4]
```diff
@@ -1390,17 +1390,24 @@ def check_for_ordered(self, op):
...
+        # TODO(PY2): use correct signature
+        # We have to do *args, **kwargs to avoid a a py2-only signature
+        # issue since np.argsort differs from argsort.
+        """Return the indicies that would sort the Categorical.
+
+        Parameters
+        ----------
```
Comment (jreback) [inline pandas/core/arrays/categorical.py]: do these doc-strings meet the new standards?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (shoyer) [reply pandas/core/arrays/base.py]: We don't necessarily want to convert into extension arrays into the same NumPy array used for `np.asarray()`.

For example, the IP Address extension array probably wants to convert into numpy object array of IPAddress object. But for sorting, it could just return numpy structured array with a few integer fields, which will be much faster for comparisons than Python IPAddress objects.

[the remark this answers is not in the document]
Comment (TomAugspurger) [reply pandas/core/arrays/base.py]: Indeed. In the IPArray case, there's a numpy array, so `_values_for_argsort` is just that array.

>  OR simply call np.asarray() if you actually need an ndarray.

Then we're back in the duplicate code situation. That's OK for the base class, but Categorical, Period, and Interval will end up re-implementing argsort from scratch. With `_values_for_argsort`, it's just a matter of constructing that array (codes, ordinals, concat left & right).

[the remark this answers is not in the document]
Comment (TomAugspurger) [reply pandas/core/arrays/base.py]: On the name `_values_for_argsort`, it's *possible* that we'll find other uses for it, in which case we just change the name.

Do you know if a simple array appropriate for arg-sorting is also appropriate for factorize, joins, indexing, etc? I'm not sure ahead of time.

File: pandas/core/arrays/categorical.py (hunk 4 above)
Comment (TomAugspurger) [reply pandas/core/arrays/categorical.py]: 7bbe796 does, aside from examples which isn't really possible.

File: pandas/core/arrays/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (TomAugspurger) [reply pandas/core/arrays/base.py]: > Do you know if a simple array appropriate for arg-sorting is also appropriate for factorize, joins, indexing, etc?

Well, we can scratch factorize / groupby off. Categorical defines `_codes_for_groupby` separately from its `ndarray_values` (codes) https://github.com/pandas-dev/pandas/blob/3783ccc694ef93c0cf159ca1ad8e69a67d828bf0/pandas/core/arrays/categorical.py#L638

So we can't just use one array for everything.

[the remark this answers is not in the document]
Comment (jreback) [reply pandas/core/arrays/base.py]: ok, so then let's pick a name change ``_codes_for_groupby``. in this refactor we want to find other usecases and fix our code now rather than later.

something like: ``_int_mapping_for_values``

**Comment by TomAugspurger** [comment]
Hmm I don't know. What about when you need a slightly different integer
mapping for values?
We already have that for Categorical, as _codes_for_groupby is
inappropriate for _values_for_argsort.

This is trying to avoid the whole "which .values is this?" issue by being
clear what it's used for. With this name, it's
very clear.

From a library author's point of view, *if* we have several of these "give
me a simple array to operate on" that can be shared by the implementation,
it's
as simple as

```
_codes_for_groupby = _values_for_argsort = _as_simple_array
```

And if they can't share an implementation (like Categorical) it's clear
which one is useful in which situation.

**Comment by TomAugspurger** [comment]
@jreback are you -1 on `_values_for_argsort`, or just -0? I believe the rest of us are +1.

**Comment by jreback** [comment]
i am -1 in this

it seems that EA is adding a bunch of things that we don’t know are really necessary or very specifically named

i would defer adding this or make it general purpose

this API has to be as simple as possible but it is exploding in size and complexity

**Comment by TomAugspurger** [comment]
> i would defer adding this or make it general purpose

I'm not sure if that's possible. We already have two cases that can't be satisfied by a simple array.

> this API has to be as simple as possible but it is exploding in size and complexity

The purpose of `_values_for_argsort` is to make it *easier* to implement `argsort` correctly. Just pass a simple array of codes if you have one (which is true for Categorical, Period, Interval, and IPArray). We need `argsort` to support groupby, which is certainly worth supporting. I think the *marginal* complexity of `_values_for_argsort` is negative. If you can't use it, then you just override `argsort` normally. If you can use it, then your job is easier.

**Comment by jorisvandenbossche** [comment]
@jreback Do you have an alternative solution? Or just not having it?

As Tom said, not having this would mean duplicating the implementation of `argsort` in `Categorical`, `Period`, `Interval` (and external extension arrays). That's certainly not ideal as well.

It is difficult to exactly know how things will turn out (whether we need to add a bunch of those `_values_for_...` methods or not, whether we can reuse this `_values_for_argsort` for other purposes or not (eg groupby or factorize), if we would need to add a constructor method for roundtripping it, ...). But it will only by actually trying to fully implement ExtensionArray and all our own arrays types (period, interval) that we will find out. 
Therefore, I personally think it is better to move forward, let's implement all those things, and when doing that, constantly re-evaluate if things can be re-used, simplified, ..

**Comment by jreback** [comment]
I think it would be worthwile to convert datetime w/tz to a first class extension type before proceeding with more changes. The issue is have is we are looking at a very narrow case, categorical, so its pretty hard to tell if something is generally needed/useful, or is something specific to an extension type. The types that are added (decimal/json) are not battle tested and don't have nearly the sophistication and internal usage as do categorical and datetime w/tz.

**Comment by TomAugspurger** [comment]
I'm close to having groupby done. I'd prefer to finish that first before starting on another type.

**Comment by TomAugspurger** [comment]
FYI @jbrockmendel has started on DatetimeArray in https://github.com/pandas-dev/pandas/pull/19902 (will review today, sorry).

**Comment by jreback** [comment]
needs a rebase. my point from above is also that the idea of mapping an EA to integers (which are meaningful) is generally useful, essentially what you have for ``_values_for_argsort`` is pretty much factorize. so maybe make this explicit, e.g. ``_factorized_values`` or something like this. (and depending on the array you *may* want to cache this).

**Comment by TomAugspurger** [comment]
The name "factorized_values" takes on pandas' factorization semantics
around types and ordering.

We can't respect the types because we don't have a good way of getting
integers (yet, and even if we did that may be expensive).
We can't respect the ordering rules (where order of appearance matters)
because that would break sorting.

What are all the properties we need, and what's a good name to summarize
those? Hard to say in the abstract.

For sorting, we need to preserve ordering and have each value be distinct.
In other words, a strictly monotonic function:
https://en.wikipedia.org/wiki/Monotonic_function.

**Comment by TomAugspurger** [comment]
Updated the name to `_simple_ndarray`, and documented the strict monotonicity requirement. I think monotonicity is a good compromise between being useful enough for a broad class of tasks, while being specific enough to be easy to implement.

**Comment by TomAugspurger** [comment]
I'm not tied to the name `_simple_ndarray`, other than

- I don't want to mention a type like integer (not all EAs can be easily converted to integers)
- I'd like to have `ndarary` in the name

**Comment by TomAugspurger** [comment]
@shoyer @jorisvandenbossche any thoughts on the name? Does strict monotonicity seem like a good requirement for this type of thing? It covers our current use-cases well I think (Categorical, Period, Interval, IPArray, datetime w/ TZ)

**Comment by jorisvandenbossche** [comment]
> It covers our current use-cases well I think

But only for sorting? Or do you see other use-cases?

If it is only for sorting, I still think including that notion in the name is better / more explicit (for now).

> [jreback] what you have for _values_for_argsort is pretty much factorize. so maybe make this explicit, e.g. _factorized_values or something like this.

I was also thinking about this at a certain point, in light of geopandas. For geopandas, I could convert the geometries in a meaningful *physical* array by converting it a (well-known) binary representation. And using this array, I could do a `factorize` step. So somehow, having a `_factorize_values` which is then used by a default implementation of `argsort` and `factorize` would make it easier for geopandas. 
However, the problem with such `_factorize_values` is that they then need to be roundtrippable, you need to convert the unique physical array back to actual ExtensionArray. In the above case of geopandas, that is possible, but I don't think that is something we can assume in general.

**Comment by TomAugspurger** [comment]
> However, the problem with such _factorize_values is that they then need to be roundtrippable

Indeed, though with a caveat, they need to be round-trippable *when combined with the original array*.

Using `PeriodIndex` as an example, we have the ordinals.

```python
In [25]: p = pd.period_range('2017', periods=4)

In [26]: labels, uniques = pd.factorize(p._ndarray_values)

In [27]: pd.PeriodIndex._from_ordinals(y, freq=p.freq)
Out[27]: PeriodIndex(['2017-01-01', '2017-01-02', '2017-01-03', '2017-01-04'], dtype='period[D]', freq='D')
```

So we would have to expand the API to include a new constructor `_from_simple_ndarray` or something like that. Is that an unreasonable burden? Keep in mind that the default implementation for the simple ndarray will still be `.astype(object)`, and the simple constructor can be aliased to `_from_scalars`, or whatever we called it.

At this point though, we're straying a bit far afield from the original purpose of this PR.

**Comment by TomAugspurger** [comment]
And, in case it wasn't clear, `.argsort` doesn't need the new constructor since we aren't reconstructing an EA from the transformed values.

**Comment by TomAugspurger** [comment]
So here's my current proposal.

    @classmethod
    def _from_simple_ndarray(self, values, similar):
        raise AbstractMethodError

    def _as_simple_ndarray(self):
        # type: () -> ndarray
        """Convert the array to a simple ndarray representaiton.

        Requirements:

        * strictly monotonic
          For every i, j index in `self`, if `self[i] < self[j]` then
          ``self._as_simple_ndarray()[i] < self._as_simple_ndarray()[j])``.
          Likewise for equality.
        * Invertable, when combined with another instance of this object
          For r = ``self._as_simple_ndarray()``,
              ``(self == self._from_simple_ndarray(r)).all()``

        """
        # Implemnetor note: This method is currently used in
        # - ExtensionArray.argsort

        return np.array(self)
```

**Comment by shoyer** [comment]
I'm pretty sure factorize can't handle arbitrary dtypes -- we only have hashtables for handful of builtin types (e.g., int64, float64, object, maybe uint64 if I recall correctly). So it seems quite likely that we will need different arrays for factorization and sorting (e.g., for IPAddress, which can return a structured array for sorting and an object array for hashing). In other cases (geopandas?) sorting might not even be well defined.

I don't like the name `_as_simple_ndarray`. It's too vague, and will both confuse users and tempt us (pandas devs) to add additional functionality/requirements to the method.

I actually like defining separate special methods such as `_ndarray_for_argsort` and `_ndarray_for_factorize`,  along with complementary constructors as necessary, e.g., `_from_factorized`. Here's why:
- The name of the methods makes it obvious (both to us and users) what properties the returned values should satisfy.
- We can define separate methods as necessary if there is any conceivable difference in use-cases. It's much easier to consolidate methods later than to split them later.
- I don't think the list of special methods like this that we will need will be very long.

Finally, for the record I'm still not a fan of putting *implementations* like the `argsort()` method on ExtensionArray. I think this is unnecessary and will make the interface more confusing. Ideally ExtensionArray should be entirely abstract.

**Comment by TomAugspurger** [comment]
> So it seems quite likely that we will need different arrays for factorization and sorting

Ah, yes. That is indeed the case for IPArray...

> Finally, for the record I'm still not a fan of putting implementations like the argsort() method on ExtensionArray.

What's the alternative? An internal helper in pandas like `argsort_array`?

**Comment by jbrockmendel** [comment]
I haven't formed an opinion on _values_for_argsort, but _conditional_ on it being needed, I'm +1 on it being kept out of the DataFrame/Series/Index namespace.

Is there an implicit assumption that EAs are always 1D?

**Comment by jorisvandenbossche** [comment]
> Finally, for the record I'm still not a fan of putting implementations like the argsort() method on ExtensionArray. I think this is unnecessary and will make the interface more confusing. Ideally ExtensionArray should be entirely abstract.

@shoyer would you then propose to have a full abstract class, and separate a partially filled in base extension array people can subclass? 
As having this "partially implemented" class to subclass is certainly useful to avoid repetition for extension authors.

**Comment by TomAugspurger** [comment]
Updated https://github.com/pandas-dev/pandas/pull/20361 with an example of what `_factorize` looks like using a similar approach (`_values_for_factorize` and a similar method for reconstructing).

**Comment by shoyer** [comment]
> @shoyer would you then propose to have a full abstract class, and separate a partially filled in base extension array people can subclass?
As having this "partially implemented" class to subclass is certainly useful to avoid repetition for extension authors.

No, it's probably not worth a separate class to inherit. I would just try to keep as much implementation away from ExtensionArray as possible.

`argsort()` really is a small wrapper around `_values_for_argsort()`, and I'd rather keep that wrapper elsewhere in pandas if possible. If we really do want to enable implementing `argsort` directly instead or (or in addition to?) `_values_for_argsort` then I suppose we do need it, but having two ways to do the same thing is a little confusing.

**Comment by jorisvandenbossche** [comment]
> but having two ways to do the same thing is a little confusing.

I think Tom did that on purpose, in the sense of: provide `_values_for_argsort` to override if you can, but if your type of data does not allow it, keep the possibility to fully override `argsort` itself.

**Comment by TomAugspurger** [comment]
> I think Tom did that on purpose

Correct. For the two layers of overriding the method (just the data being sorted, or the entire implementation).

**Comment by shoyer** [comment]
> I think Tom did that on purpose, in the sense of: provide _values_for_argsort to override if you can, but if your type of data does not allow it, keep the possibility to fully override argsort itself.

Are there any use cases where we would need to fully implement `argsort()`?

**Comment by TomAugspurger** [comment]
No concrete use-cases yet. I could imagine someone wanting to use their own sorting algorithm, but perhaps we shouldn't design around speculations.

**Comment by shoyer** [comment]
> No concrete use-cases yet. I could imagine someone wanting to use their own sorting algorithm, but perhaps we shouldn't design around speculations.

This doesn't feel very realistic to me, unless we're imagining the ability to swap-out pandas's entire computational machinery (which seems out of scope for ExtensionArray). Pandas/NumPy already have highly performant sorting algorithms.

**Comment by TomAugspurger** [comment]
OK, in that case, are we OK with ExtensionArray having an argsort method to
ease the use within pandas? Having argsort means
we don't need to worry about whether we have an ExtensionArray or an
ndarray.

**Comment by shoyer** [comment]
> OK, in that case, are we OK with ExtensionArray having an argsort method to
ease the use within pandas? Having argsort means
we don't need to worry about whether we have an ExtensionArray or an
ndarray.

I would rather write a internal function for that and/or add an ExtensionArray wrapper that is used internally.

**Comment by jreback** [comment]
> I would rather write a internal function for that and/or add an ExtensionArray wrapper that is used internally.

-1 on this. we are adding even more boilerplate / complexity here. Let's just keep this simple. having impl in EA is just fine to be honest.

**Comment by shoyer** [comment]
I don't see the downside of defining a separate internal module that we use for function calls on ndarray/ExtensionArray objects:
```
def argsort(values, ...):
    if isinstance(values, np.ndarray):
        return values.argsort(...)
    elif isinstance(values, ExtensionArray):
        codes = values._values_for_argsort()
        ...
    else:
        raise TypeError
```

There's no inherent advantage to using methods for everything.

If desired, we could even use something like `functools.singledispatch` but that's probably overkill here.

**Comment by jreback** [comment]
> There's no inherent advantage to using methods for everything.

sure there is. reduced code complexity and DRY. The problem is no-one can find things. co-locating with in a single place is very helpful.

**Comment by shoyer** [comment]
I agree that keeping things in one-place has value. But I think it's much more valuable to have a clear interface for our users.

**Comment by jreback** [comment]
I agree is best to have a single interface for users. To be honest I would let the sub-classes just re-define ``.argsort()`` if needed and dispense with ``_values_for_argsort``, otherwise we end up having serveral of these little 'interfaces.

**Comment by TomAugspurger** [comment]
Well, it looks like I agree with no one, except perhaps Joris :) So let's see if we can get this unstuck.

@shoyer favors just `_values_for_argsort`, `_values_for_factorize`, with no `.argsort`, and `.factorize` methods. The goal being a *clear* interface for users. FWIW, I care less about the clarity of the interface than the *ease of implementation* (users first, pandas second). For user's ease of implementation, both are similar, and I think superior to just providing `.argsort`, with no `_values_for_argsort`. For pandas, having both is easiest as it lets us write code that can work on either ndarrays, Series, Indexes, or EAs.

One point in @shoyer's favor is "why do we have `.argsort` and not `.sort_values`?" The answer that `.argsort` is all that's needed for `NDFrame.sort_values`, groupby, etc. Not exactly the clearest.

> [@jreback] otherwise we end up having serveral of these little 'interfaces.

I think that's a good thing. For `.argsort`, things are relatively easy to do (just have to correctly handle `ascending`). For `.factorize` things are much trickier. I don't think every author should have to reimplement that logic.

I feel like I've written a lot of text and not gotten any closer to a solution :/ I think that's because I don't have a feel for how difficult an EA wrapper (class or just functions) would be. I'll explore that today.

**Comment by jorisvandenbossche** [comment]
Many of things we are discussing here, we will only encounter in practice (like the question if we will get many of such `_values_for_argsort` interfaces). So I think it is more important to get something minimally working merged, so people (and we in pandas) can start using and experimenting with it. 

I am fine with the current PR (both `_values_for_argsort` and `argsort`), although I personally don't really care much (from a user perspective) for having `argsort`. 
So a "minimal" conservative route to go forward (minimal from user point of view), would also be to only have `_values_for_argsort` for extension authors. 
We can always later add more public methods if there is demand for it. And I think it is also no problem to still change things in the extension interface (for the extension author, not user) later on (I suppose we will label this experimental for some time).

Although having both `_values_for_argsort` and `argsort` methods on ExtensionArray might actually be kind of a compromise between the viewpoints of Jeff and Stephan :-)

**Comment by shoyer** [comment]
> So I think it is more important to get something minimally working merged, so people (and we in pandas) can start using and experimenting with it.

👍

**Comment by TomAugspurger** [comment]
I think (hope?) then that we're OK proceeding with this as is then.

https://github.com/pandas-dev/pandas/pull/19957/commits/c776133d42beb81d94ab5000da8bb2ad56f53e90 unskips the sorting tests for JSONArray. The strategy is to sort an array of the dictionary items converted to tuples, which is maybe the most sensible way to sort a dictionary. Anyway, we have an example of sorting something that isn't regularly sortable.

The remaining skip is for `df.sort_values(['A', 'B'])`, which factorizes the column. That should be fixed by https://github.com/pandas-dev/pandas/pull/20361, which this is blocking.

Then we'll have groupby, which is shaping up to be a surprisingly small change.

**Comment by TomAugspurger** [comment]
CI all passed.

**Comment by TomAugspurger** [comment]
@jreback are you +1 here, given the discussion above? Since the last time around we discovered that `_ndarray_values` can't hope to be used for both factorize and argsorting.

**Comment by TomAugspurger** [comment]
@jreback thoughts?

**Comment by jreback** [comment]
will look a bit later

**Review by jreback — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
> Since its completely private

Note that those APIs are *not* private. They are part of the extension array interface, so 'public' for extension authors.

> there is a fair amount of divergence between naming inside pandas internals and EA

First, I am not sure this would be a bad thing. EAs are *new*. If we go the route of Series and Index being composed of an ExtensionArray (and not subclassing them), then they are something different, and I think it is good to use different names for things that do something differently. 

But, I currently also don't see any divergence.
The specific methods we now have for ExtensionArrays:

- `_constructor_from_sequence` -> the same concept does not exist for Series/Index, so new name is normal
- `_values_for_argsort` -> this is not relevant for Series/Index, something new for EA
- `_formatting_values` -> this is consistent with the Block method
- `_concat_same_type` -> this is consistent with the Block method (and there is also no equivalent for Series). There is a `Index._concat_same_dtype`, which is maybe something to investigate if we can get rid of that.
- `_can_hold_na` -> this is consistent with both Series and Block method
- `_ndarray_values` -> consistent with naming of equivalent attribute on Series/Index

It is of course true that there is divergence with numpy arrays, but if that was not needed, we wouldn't have needed to make such an ExtensionArray in the first place.

**Comment by jreback** [comment]
https://circleci.com/gh/pandas-dev/pandas/12793#tests/containers/3

is breaking. I think some json EA tests are depending on a certain ordering. 

@TomAugspurger

**Comment by jreback** [comment]
@jorisvandenbossche 

> First, I am not sure this would be a bad thing. EAs are new. If we go the route of Series and Index being composed of an ExtensionArray (and not subclassing them), then they are something different, and I think it is good to use different names for things that do something differently.

My point is there are now 2 ways / names of doing things (in some cases). This is not great and should be addressed by re-aligning naming in the Series. 

A new reader to the codebase will be very confused on what to use when. Sure EA's are not exactly like Series/Index. But over the years we have gone to great lengths to make these look and feel and be implemented in a very similar way. So the divergence between Index / EA's is disturbing. My point is that before we move more on EA, Index's need to be subclassed and become real EA's.

**Comment by jorisvandenbossche** [comment]
> My point is there are now 2 ways / names of doing things (in some cases).

Which cases? Please be specific. I gave a list above, and I don't see any with 2 ways of naming.

**Comment by jreback** [comment]
Index is not an EA subclass.

**Comment by jorisvandenbossche** [comment]
Sorry, what is the relation with my previous comment? 
Yes, Index is not a EA subclass, and that is on purpose because 1) that's a much bigger undertaking 2) I don't think we already agree on whether it *should* become one or if we should do composition (as we do with normal arrays)

**Comment by jreback** [comment]
my point is that when Index becomes a subclass that all of the implementation details of it should follow a single convention (for construction / indexing) etc. These would naturally use the EA conventions. The problem now is they are different. This is just technical debt.

**Comment by jorisvandenbossche** [comment]
@jreback I am not sure that Index will become a subclass of ExtensionArray, but let's not discuss that here, but rather in https://github.com/pandas-dev/pandas/issues/19696#issuecomment-375653047
